# Unemployment Analysis - PLFS 2023-24 (India)

Dataset - Periodic Labour Force Survey(PLFS), 2023-24
Source - Ministry of Statistics and Programme Implementation (MoSPI), Government of India
Indicators - Unemployment Rate (UR), Worker Population Ratio (WPR), Labour Force Participation Rate (LFPR)
Age Group - 15-59 years
Coverage - 36 States/UTs + All India Benchmark

## Objective

This notebook performs **data loading, inspection, cleaning, and preparation** of the PLFS 2023-24 dataset.  
The cleaned data is split into two separate datasets:
- `state_data.csv` — State/UT level data for analysis
- `national_data.csv` — All India benchmark data for comparison

## Step 1: Import Libraries
Importing important libraries which will be required for data loading, inspection, cleaning, exporting and visualization. 

Note:- In this notebook only pandas is imported as the work is mainly related to loading, cleaning and exporting. 

In [1]:
# Import the library
import pandas as pd
print("Library Imported")

Library Imported


## Step 2: Load dataset
Loading the PLFS dataset from the data/ folder.

In [2]:
# Load the dataset from excel file
df = pd.read_excel("../data/raw/unemployment_analysis.xlsx")
print(" Dataset Loaded Successfully")

# Understanding the columns and rows of the datset
print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")

# Preview head
df.head()

 Dataset Loaded Successfully
Shape: 999 rows × 7 columns


,year,state,sector,gender,indicator,value,level
0,2023-24,Andhra Pradesh,Rural,Male,UR,3.8,State
1,2023-24,Arunachal Pradesh,Rural,Male,UR,5.4,State
2,2023-24,Assam,Rural,Male,UR,3.6,State
3,2023-24,Bihar,Rural,Male,UR,3.3,State
4,2023-24,Chhattisgarh,Rural,Male,UR,1.6,State


## Step 3: Data Inspection
Understanding the data and finding any missing values, data types and unique values in each columns.

### Step 3.1: Dataset info
Checking columns name, data types and non-null values.

In [3]:
#  Overview of dataset — dtypes, non-null counts, memory usage
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 999 entries, 0 to 998
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   year       999 non-null    object 
 1   state      999 non-null    object 
 2   sector     999 non-null    object 
 3   gender     999 non-null    object 
 4   indicator  999 non-null    object 
 5   value      990 non-null    float64
 6   level      999 non-null    object 
dtypes: float64(1), object(6)
memory usage: 54.8+ KB


### Step 3.2: Descriptive Analysis

Performing initial data exploration to understand:
- Structure of dataset
- Distribution of categorical variables
- Summary statistics of numerical values

In [4]:
# Statistical summary of all columns
# include='all' shows stats for both numeric and text columns
df.describe(include='all')

,year,state,sector,gender,indicator,value,level
count,999,999,999,999,999,990.000000,999
unique,1,37,3,3,3,NaN,2
top,2023-24,Andhra Pradesh,Rural,Male,UR,NaN,State
freq,999,27,333,333,333,NaN,972
mean,NaN,NaN,NaN,NaN,NaN,40.853232,NaN
std,NaN,NaN,NaN,NaN,NaN,29.070807,NaN
min,NaN,NaN,NaN,NaN,NaN,0.000000,NaN
25%,NaN,NaN,NaN,NaN,NaN,6.725000,NaN
50%,NaN,NaN,NaN,NaN,NaN,47.400000,NaN
75%,NaN,NaN,NaN,NaN,NaN,68.375000,NaN


### Step 3.3: Column Names

In [5]:
# List all column names in the dataset
print("Columns", list(df.columns))

Columns ['year', 'state', 'sector', 'gender', 'indicator', 'value', 'level']


### Step 3.4: Missing Values
Checking null counts per column.

In [6]:
# Count missing values per column
print("Missing values per column:")
print(df.isnull().sum())

# Show exactly which rows have missing values
print("\nRows with missing values:")
print(df[df['value'].isnull()][['state', 'sector', 'gender', 'indicator']])

Missing values per column:
year         0
state        0
sector       0
gender       0
indicator    0
value        9
level        0
dtype: int64

Rows with missing values:
          state sector  gender indicator
30   Chandigarh  Rural    Male        UR
67   Chandigarh  Rural  Female        UR
104  Chandigarh  Rural  Person        UR
363  Chandigarh  Rural    Male       WPR
400  Chandigarh  Rural  Female       WPR
437  Chandigarh  Rural  Person       WPR
696  Chandigarh  Rural    Male      LFPR
733  Chandigarh  Rural  Female      LFPR
770  Chandigarh  Rural  Person      LFPR


### Step 3.5: Handling Structural Missing and Duplicate Values

Chandigarh has no rural population as per PLFS survey design.
Therefore, rows corresponding to Chandigarh (Rural sector) are removed,
as they are not applicable for analysis.

In [7]:
df_original = df.copy()

In [8]:
# Remove missing values 
df_cleaned = df.dropna(subset=['value']).copy()

In [9]:
# Checking the dataset shape for reducing any error
print("Original shape:", df.shape)
print("Cleaned shape:", df_cleaned.shape)

Original shape: (999, 7)
Cleaned shape: (990, 7)


In [10]:
# checking for duplicate rows
print("Duplicate rows after cleaning:", df_cleaned.duplicated().sum())

Duplicate rows after cleaning: 0


### Step 3.6: Category Distribution

Checking row count for each category to verify the data is balanced and complete.

- Sector Distribution Insight
Urban and Rural+Urban categories have equal representation (333 each),
while Rural has slightly fewer observations (324) due to removal of Chandigarh rural data.

In [11]:
# Distribution of indicators — UR, WPR, LFPR
print("Indicator distribution:")
print(df_cleaned['indicator'].value_counts())

Indicator distribution:
indicator
UR      330
WPR     330
LFPR    330
Name: count, dtype: int64


In [12]:
# Distribution of sectors — Rural, Urban, Rural+Urban
# Note: Rural has fewer rows due to 9 missing Chandigarh Rural rows
print("Sector distribution:")
print(df_cleaned['sector'].value_counts())

Sector distribution:
sector
Urban          333
Rural+Urban    333
Rural          324
Name: count, dtype: int64


In [13]:
# Distribution of gender — Male, Female, Person (combined)
print("Gender distribution:")
print(df_cleaned['gender'].value_counts())

Gender distribution:
gender
Male      330
Female    330
Person    330
Name: count, dtype: int64


In [14]:
# Distribution of level — State vs National (All India benchmark)
print("Level distribution:")
print(df_cleaned['level'].value_counts())

Level distribution:
level
State       963
National     27
Name: count, dtype: int64


## Step 4: Rename Indicators and verify year

- Renaming indicator abbreviations to descriptive names for better understanding.
- Confirming this is single year dataset (2023-24).

In [15]:
# Renaming indicator names 
df_cleaned['indicator'] = df_cleaned['indicator'].replace({
    'UR': 'Unemployment Rate',
    'WPR': 'Worker Population Ratio',
    'LFPR': 'Labour Force Participation Rate'
})

In [16]:
# Confirming only one survey year exist
print(f"Year in dataset: {df_cleaned['year'].unique()}")

Year in dataset: ['2023-24']


## Step 5: Split into State and National Datasets

The dataset contains two types of records 
- State - Data for individual state/UTs
- National - All India aggregate data(all India row )

The split is necessary to make datasets reliable. 
If both results are kept together and compute averages then all India row get mixed in state leads to wrong outcomes.


In [17]:
# Split datasets into state and national datasets
df_state = df_cleaned[df_cleaned['level'] == 'State'].copy()
df_india = df_cleaned[df_cleaned['level'] == 'National'].copy()

In [18]:
# Drop 'level' column after splitting
df_state = df_state.drop(columns=['level'])
df_india = df_india.drop(columns=['level'])

In [19]:
# Keep only relevant columns
cols = ['year', 'state', 'sector', 'gender', 'indicator', 'value']

df_state = df_state[cols]
df_india = df_india[cols]

### Step 5.1: Preview both datasets

In [20]:
# Preview first 5 rows of state dataset
print("STATE DATASET:")
df_state.head(5)

STATE DATASET:


,year,state,sector,gender,indicator,value
0,2023-24,Andhra Pradesh,Rural,Male,Unemployment Rate,3.8
1,2023-24,Arunachal Pradesh,Rural,Male,Unemployment Rate,5.4
2,2023-24,Assam,Rural,Male,Unemployment Rate,3.6
3,2023-24,Bihar,Rural,Male,Unemployment Rate,3.3
4,2023-24,Chhattisgarh,Rural,Male,Unemployment Rate,1.6


In [21]:
# Preview national dataset
print("NATIONAL DATASET:")
df_india

NATIONAL DATASET:


,year,state,sector,gender,indicator,value
36,2023-24,all India,Rural,Male,Unemployment Rate,2.7
73,2023-24,all India,Rural,Female,Unemployment Rate,2.1
110,2023-24,all India,Rural,Person,Unemployment Rate,2.5
147,2023-24,all India,Urban,Male,Unemployment Rate,4.4
184,2023-24,all India,Urban,Female,Unemployment Rate,7.1
221,2023-24,all India,Urban,Person,Unemployment Rate,5.1
258,2023-24,all India,Rural+Urban,Male,Unemployment Rate,3.2
295,2023-24,all India,Rural+Urban,Female,Unemployment Rate,3.2
332,2023-24,all India,Rural+Urban,Person,Unemployment Rate,3.2
369,2023-24,all India,Rural,Male,Worker Population Ratio,78.1


## Step 6: Data Validation Checks (Post-Cleaning)

Performing final validation checks to ensure:
- Data consistency after cleaning
- No unintended data loss
- Structural completeness of dataset

In [22]:
# Validating dataset structure by checking total records and unique categories 
# (states, indicators, and sectors) after cleaning
print("Final cleaned dataset shape:", df_cleaned.shape)
print("Unique states:", df_cleaned['state'].nunique())
print("Unique indicators:", df_cleaned['indicator'].unique())
print("Unique sectors:", df_cleaned['sector'].unique())

Final cleaned dataset shape: (990, 7)
Unique states: 37
Unique indicators: ['Unemployment Rate' 'Worker Population Ratio'
 'Labour Force Participation Rate']
Unique sectors: ['Rural' 'Urban' 'Rural+Urban']


In [24]:
# Validating that all combinations of indicator, gender, and sector are present 
# and have consistent record counts
print("Check combinations:")
print(df_cleaned.groupby(['indicator', 'gender', 'sector']).size().head(10))

Check combinations:
indicator                        gender  sector     
Labour Force Participation Rate  Female  Rural          36
                                         Rural+Urban    37
                                         Urban          37
                                 Male    Rural          36
                                         Rural+Urban    37
                                         Urban          37
                                 Person  Rural          36
                                         Rural+Urban    37
                                         Urban          37
Unemployment Rate                Female  Rural          36
dtype: int64


In [25]:
# Sorting the dataset by state, indicator, gender, and sector to maintain a consistent and logical order
df_cleaned = df_cleaned.sort_values(by=['state', 'indicator', 'gender', 'sector'])

In [26]:
# Resetting index to ensure a clean and continuous row numbering after data cleaning
df_cleaned = df_cleaned.reset_index(drop=True)

## Step 7: Export Cleaned Dataset 

In [27]:
# Exported clean dataset
df_cleaned.to_csv("../data/processed/cleaned.csv", index=False)

# State dataset
df_state.to_csv("../data/processed/state_data.csv", index=False)
print("State Data CSV Saved Successfully")
 
# National Dataset
df_india.to_csv("../data/processed/national_data.csv", index=False)
print("National Data CSV Saved Successfully")

State Data CSV Saved Successfully
National Data CSV Saved Successfully


## Data Cleaning Summary

- Initial dataset: 999 rows
- Identified 9 missing values in 'value' column (Chandigarh Rural)
- These rows are structurally invalid (no rural population in Chandigarh)
- Removed 9 rows → final dataset: 990 rows
- No duplicate records found
- Indicators renamed for clarity
- Dataset split into:
  - State-level dataset
  - National benchmark dataset